ENH: Add typing for RandomState

bashtage opened the pull request:
Add typing for RandomState
Small fixes to Generator's typing

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

bashtage committed:
ENH: Add typing for RandomState

Add typing for RandomState
Small fixes to Generator's typing

bashtage commented on numpy/random/mtrand.pyi:
```diff
@@ -0,0 +1,551 @@
+
+_rand: RandomState
+
+beta = _rand.beta
```
Should I define these here? Probably, but it make a bit of a mess in `__init__.pyi`

BvB93 replied:
Yeah, they should definitely be in here.

As for `__init__.pyi`, you could try importing try importing multiple objects with a single `import` statement, that might look a bit neater:

``` python
from numpy.random.mtrand import (
    RandomState as RandomState,
    beta as beta,
    binomial as binomial,
    ...  # etc.
)
```

BvB93 commented:
Closes https://github.com/numpy/numpy/issues/17103.

BvB93 commented:
I think that it would be wise to add a few typing tests for `np.random` at this point.

There are currently three types of typing tests: `reveal`, `pass` and `fail`:
* The `reveal` tests check whether the observed mypy reveal is indeed what one would expect ([example](https://github.com/numpy/numpy/blob/master/numpy/typing/tests/data/reveal/arithmetic.py)).
* The `pass` tests check whether the code passes without any errors, be it errors produced by mypy or by actually executing the code during runtime ([example](https://github.com/numpy/numpy/blob/master/numpy/typing/tests/data/pass/arithmetic.py)).
* The `fail` tests check that faulty expressions return the mypy error that one might expect of them ([example](https://github.com/numpy/numpy/blob/master/numpy/typing/tests/data/fail/arithmetic.py)).

bashtage committed:
MAINT: Restructure imports in pyi file

bashtage committed:
TST: Add tests for random typing

bashtage committed:
ENH/BUG: Add tests and correct typing issues

Add tests for lower level components
Correct errors found in testing
Improve specificity of tests
Other small fixes to docs and typing

bashtage commented on numpy/typing/tests/test_typing.py:
```diff
@@ -48,8 +48,11 @@ def run_mypy() -> None:
+    NUMPY_TYPING_TEST_CLEAR_CACHE=0 pytest numpy/typing/tests
     """
-    if os.path.isdir(CACHE_DIR):
+    if os.path.isdir(CACHE_DIR) and bool(os.environ.get("NUMPY_TYPING_TEST_CLEAR_CACHE", True)):
```
@BvB93 I added this to speed up pytest runs. Seems pretty harmless but wanted to draw your attention to it.

BvB93 replied:
This should be fine I suppose, just be aware that a stale mypy cache can potentially causes unexpected results.

bashtage committed:
ENH/BUG: Improve definitions and fix bugs

Add tests
Fix related bugs
Improve specificity of definitions

BvB93 commented:
At first glance this PR looks just fine; I'll give it a final once-over later this week.

bashtage committed:
TST: Add tests for dtype and out

Add further tests for functions that accept out and dtype arguments

bashtage committed:
BUG: Corect typing statements in mtrand

Bring over changes from Generator to RandomState where appropriate
Remove attributes not in RandomState

bashtage committed:
ENH: Add tests for RandomState

bashtage committed:
BUG: Fix typing bugs

bashtage committed:
Small fixes

bashtage commented:
@BvB93 I'm struggling with functions that have outputs `ndarray[Any, type[int_]]` since `int_` has different sizes on different platforms.  Can these be typed checked using `reveal`?  Or do I need to write different tests for Windows and Linux? Or something else?

Commit:
Use {int_} for int return types

bashtage committed:
TST: Add final tests

bashtage commented:
@BvB93 I found your typing template in the tests, e.g. `{int_}`.

BvB93 commented on numpy/random/mtrand.pyi:
```diff
@@ -0,0 +1,588 @@
+    def __str__(self) -> str: ...
+    def __getstate__(self) -> Dict[str, Any]: ...
+    def __setstate__(self, state: Dict[str, Any]) -> None: ...
+    def __reduce__(self) -> Tuple[Callable[[str], BitGenerator], Tuple[str], Dict[str, Any]]: ...
```
```suggestion
    def __reduce__(self) -> Tuple[Callable[[str], RandomState], Tuple[str], Dict[str, Any]]: ...
```
I assume this should return a `RandomState` instance?

bashtage replied:
Yes, thanks.

BvB93 commented on numpy/random/mtrand.pyi:
```diff
@@ -0,0 +1,588 @@
+        low: int,
+        high: Optional[int] = ...,
+    ) -> int: ...
+    @overload
+    def randint(  # type: ignore[misc]
+        self,
+        low: int,
+        high: Optional[int] = ...,
+        size: None = ...,
+        dtype: _DTypeLikeBool = ...,
+        endpoint: bool = ...,
+    ) -> bool: ...
+    @overload
+    def randint(  # type: ignore[misc]
+        self,
+        low: int,
+        high: Optional[int] = ...,
+        size: None = ...,
+        dtype: _DTypeLikeBool = ...,
+    ) -> bool: ...
+    @overload
+    def randint(  # type: ignore[misc]
+        self,
+        low: int,
+        high: Optional[int] = ...,
+        size: None = ...,
+        dtype: Union[_DTypeLikeInt, _DTypeLikeUInt] = ...,
+    ) -> int: ...
```
Uhm, so what's going on here with these three completely different signatures?

bashtage replied:
I couldn't get it to handle the default values properly.  I read around on the mypy  github about complex defaults and this is what I came up with.  Probably no minimal.

bashtage replied:
I've been able to clean this up a lot.

BvB93 replied:
I just realized I selected the wrong lines; it's really these first two overloads that I don't understand (_e.g._ what's this `endpoint` parameter doing here all of a sudden?).
``` python
    @overload
    def randint(  # type: ignore[misc]
        self,
        low: int,
        high: Optional[int] = ...,
    ) -> int: ...
    @overload
    def randint(  # type: ignore[misc]
        self,
        low: int,
        high: Optional[int] = ...,
        size: None = ...,
        dtype: _DTypeLikeBool = ...,
        endpoint: bool = ...,
    ) -> bool: ...
```

bashtage replied:
I need the default path to come first since this produces an `int`, then the more specific `bool` path, then the full `int` path.

I followed what I learned from 

https://github.com/python/mypy/issues/5486

when going with this strategy. I think the problem is that a bool is an int but not vice versa. If I omit the short def, and then have the `int` def first, then the scalar `bool` is reported as an `int`, which isn't totally wrong.  This strategy allowed me to get the most specific type information.

BvB93 commented on numpy/random/mtrand.pyi:
```diff
@@ -0,0 +1,588 @@
+        size: Optional[_ShapeLike] = ...,
+    ) -> ndarray[Any, dtype[float64]]: ...
+    @overload
+    def rand(self, *args: None) -> float: ...
```
```suggestion
    def rand(self) -> float: ...
```
`*args: None` implies that `np.random.rand(None)` should be a valid input.

bashtage replied:
Thanks.

BvB93 commented on numpy/random/mtrand.pyi:
```diff
@@ -0,0 +1,588 @@
+    @overload
+    def rand(self, *args: int) -> ndarray[Any, dtype[float64]]: ...
+    @overload
+    def randn(self, *args: None) -> float: ...
```
```suggestion
    def randn(self) -> float: ...
```
Same comment as for `rand`.

BvB93 reviewed (commented):
@bashtage it took a bit longer than expected but I've made some time to go through the PR.
There are a few comments but overall I'd say it's looking pretty good.

BvB93 commented:
> @BvB93 I found your typing template in the tests, e.g. `{int_}`.

If you ever stumble upon a common pattern that you feel might be useful to have as an alias, 
then go ahead add it to the dictionary.

bashtage commented on numpy/typing/tests/data/fail/random.py:
```diff
@@ -0,0 +1,61 @@
+from typing import Any, List
+
+SEED_FLOAT: float = 457.3
+SEED_ARR_FLOAT: np.ndarray[Any, np.dtype[np.float64]] = np.array([1.0, 2, 3, 4])
```
I went with the belt and bracers approach for now.

bashtage commented on numpy/random/_generator.pyi:
```diff
@@ -92,17 +87,31 @@ class Generator:
+        self,
+        *,
+        out: ndarray[Any, dtype[float64]] = ...,
+    ) -> ndarray[Any, dtype[float64]]: ...
```
Is that OK here since out has to be ` ndarray[Any, dtype[float64]]` and can't be any other type?

bashtage replied:
To be specific

In these 3:

1. When using `out` and not type, `out` must be a `float64` array.
2. When using `dtype=float32` out can be `None` or a `float32` array, but nothing else.
3. When using `dtype=float64` out can be `None` or a `float64` array, but nothing else.

Can I simplify these? Because of the `dtype` keyword argument that defaults to `np.float64`, it isn't a case of getting the type from `out` when `dtype` is missing.  Would have been a good design to do this, but I think that boat has sailed.

```
    @overload
    def standard_normal(  # type: ignore[misc]
        self,
        *,
        out: ndarray[Any, dtype[float64]] = ...,
    ) -> ndarray[Any, dtype[float64]]: ...
    @overload
    def standard_normal(  # type: ignore[misc]
        self,
        size: _ShapeLike = ...,
        dtype: _DTypeLikeFloat32 = ...,
        out: Optional[ndarray[Any, dtype[float32]]] = ...,
    ) -> ndarray[Any, dtype[float32]]: ...
    @overload
    def standard_normal(  # type: ignore[misc]
        self,
        size: _ShapeLike = ...,
        dtype: _DTypeLikeFloat64 = ...,
        out: Optional[ndarray[Any, dtype[float64]]] = ...,
    ) -> ndarray[Any, dtype[float64]]: ...
```

BvB93 replied:
Considering the limited set of accepted dtypes I don't think you can actually simplify this any further.
Let's just keep things as they are.

bashtage committed:
CLN: Clean up random typing

Clean unnecessary definitions
Fix small typing errors

BvB93 commented on numpy/random/mtrand.pyi:
```diff
@@ -0,0 +1,580 @@
+        high: Optional[int] = ...,
+        size: None = ...,
+        dtype: _DTypeLikeBool = ...,
+        endpoint: bool = ...,
```
```suggestion
```
`randint` doesn't have an `endpoint` parameter
https://github.com/numpy/numpy/blob/9437402f1d60778cd723e6119044ef0510ef17c0/numpy/random/mtrand.pyx#L645

bashtage replied:
Thanks. Fixed.

bashtage committed:
BUG: Remove incorrect variable name

Remove incorrect variable name in randint

BvB93 reviewed (approved):
LGTM

mattip commented on numpy/random/__init__.pyi:
```diff
@@ -8,57 +8,58 @@ from numpy.random._philox import Philox as Philox
 from numpy.random._sfc64 import SFC64 as SFC64
 from numpy.random.bit_generator import BitGenerator as BitGenerator
 from numpy.random.bit_generator import SeedSequence as SeedSequence
+from numpy.random.mtrand import (
```
Does importing here work differently than regular python that we need to do "from x import y as y" and not "from x import y"? If so, it would be nice to get a comment explaining this for the next time I ask the same question

bashtage replied:
Taking @BvB93 's word, yes.  If you want a symbol reexported, then you need to import X as X.

BvB93 replied:
Per PEP [484](https://www.python.org/dev/peps/pep-0484/):

> Modules and variables imported into the stub are not considered exported from the stub unless the import uses the `import ... as ...` form or the equivalent `from ... import ... as ...` form. (UPDATE: To clarify, the intention here is that only names imported using the form X as X will be exported, i.e. the name before and after as must be the same.)

mattip replied:
Thanks. I probably asked this last time I looked at this code.

mattip commented:
Thanks @bashtage
